charms/tensorboards-web-app/tests/integration/test_charm.py

```diff
@@ -0,0 +1,101 @@
+
+import logging
+from pathlib import Path
+
+import aiohttp
+import pytest
+import yaml
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+CONFIG = yaml.safe_load(Path("./config.yaml").read_text())
+APP_NAME = "tensorboards-web-app"
+
+ISTIO_GATEWAY = "istio-ingressgateway"
+ISTIO_PILOT = "istio-pilot"
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest):
+    """Build and deploy the charm under test.
+    Assert on the unit status before ingress relation is set up.
+    """
+    charm = await ops_test.build_charm(".")
+    image_path = METADATA["resources"][f"{APP_NAME}-image"]["upstream-source"]
+    resources = {f"{APP_NAME}-image": image_path}
+
+    await ops_test.model.deploy(charm, resources=resources, application_name=APP_NAME, trust=True)
+
+    await ops_test.model.wait_for_idle(apps=[APP_NAME], status="blocked", timeout=60 * 5)
+
+    unit = ops_test.model.applications[APP_NAME].units[0]
+    assert unit.workload_status == "blocked"
+    assert unit.workload_status_message == "No ingress relation available"
+
+
+@pytest.mark.abort_on_fail
+async def test_ingress_relation(ops_test: OpsTest):
+    """Setup Istio and relate it to the Tensoboards Web App(TWA)."""
+    await setup_istio(ops_test, ISTIO_GATEWAY, ISTIO_PILOT)
+
+    await ops_test.model.add_relation(f"{ISTIO_PILOT}:ingress", f"{APP_NAME}:ingress")
+
+    await ops_test.model.wait_for_idle(apps=[APP_NAME], status="active", timeout=60 * 5)
+
+
+@pytest.mark.abort_on_fail
+async def test_ui_is_accessible(ops_test: OpsTest):
+    """Verify that TWA's UI is accessible."""
+    # This test also checks that Pebble checks pass since it uses the same URL.
+
+    status = await ops_test.model.get_status()
+    units = status["applications"][APP_NAME]["units"]
+    url = units[f"{APP_NAME}/0"]["address"]
+
+    # obtain status and response text from TWA URL
+    port = CONFIG["options"]["port"]["default"]
+    result_status, result_text = await fetch_response(f"http://{url}:{port}")
+
+    # verify that UI is accessible
+    assert result_status == 200
+    assert len(result_text) > 0
+    assert "Tensorboards Manager UI" in result_text
+
+
+async def setup_istio(ops_test: OpsTest, istio_gateway: str, istio_pilot: str):
+    """Deploy Istio Ingress Gateway and Istio Pilot."""
+    await ops_test.model.deploy(
+        entity_url="istio-gateway",
+        application_name=istio_gateway,
+        channel="latest/edge",
+        config={"kind": "ingress"},
+        trust=True,
+    )
+    await ops_test.model.deploy(
+        istio_pilot,
+        channel="latest/edge",
+        config={"default-gateway": "test-gateway"},
+        trust=True,
+    )
+    await ops_test.model.add_relation(istio_pilot, istio_gateway)
+
+    await ops_test.model.wait_for_idle(
+        apps=[istio_pilot, istio_gateway],
+        status="active",
+        timeout=60 * 5,
+    )
```

Comment:
In the tensorboard controller rewrite I moved this helper to a separate `utils.py` module. If we're reusing stuff though, perhaps it would be worth considering moving some of them to `chisme` as we had discussed.

Reply:
Sure, moving them to `chisme` makes total sense. I 'm not sure if we have though any test related functions in chisme ATM but this could be the start. That being said, I suggest to not block on this. We can merge this PR and in the meantime, I 'll send a PR in `chisme` adding functions from the `utils.py` file you 've introduced.

Reply:
Absolutely, let's not block on this! It'd be great to have testing utils like that in chisme, but that is indeed a separate discussion!